submission/pipelines/generate_pipelines.py

```diff
@@ -396,33 +404,171 @@ def remove_digests(
     for step in pipeline_json_structure['steps']:
         if step['primitive']['id'] not in exclude_primitives:
             del step['primitive']['digest']
+        
 
     return pipeline_json_structure
 
 
-for task_type in ['classification', 'regression']:
-    pipeline = generate_imputer_pipeline(task_type)
-    pipeline_json_structure = pipeline.to_json_structure()
-    remove_digests(
-        pipeline_json_structure, exclude_primitives={
-            RandomSamplingImputer.metadata.query()['id']
-        }
-    )
-    pipeline_path = './pipelines/random_sampling_imputer/{}.json'.format(
-        pipeline_json_structure['id']
-    )
-    json.dump(pipeline_json_structure, open(pipeline_path, 'w'), indent=4)
-    os.chmod(pipeline_path, 0o777)
+def update_digest(
+    pipeline_json_structure, filename=None
+):
+    """
+    This function updates the pipeline's digests and version numbers
+
+    Parameters
+    ----------
+    pipeline_json_structure: the pipeline in JSON form (WITHOUT) digests.  This or the `filename` parameter is mandatory
+    filename: the filename of the pipeline json, so we can read it in
+
+    :return a pipeline with updated digests
+    """
+    if pipeline_json_structure is None and filename is None:
+        raise Exception
+    elif pipeline_json_structure is None:
+        with open(filename, "r") as file:
+            # NOTE: must be a pipeline with no digests, or recent digests
+            # NOTE: reading this in as straight JSON doesn't work so we have to use the pipeline_module
+            pipeline_json_structure = pipeline_module.Pipeline.from_json(string_or_file=file).to_json_structure()
+    else:
+        try:
+            pipeline_json_structure = pipeline_module.Pipeline.from_json(json.dumps(pipeline_json_structure)).to_json_structure()
+        except Exception as e:
+            pass
+    for step in pipeline_json_structure['steps']:
+        # if not updated, check and update
+        primitive = pipeline_module.PrimitiveStep(
+            primitive=d3m_index.get_primitive(
+                step["primitive"]["python_path"]
+            )
+        )
+        check_step = primitive.to_json_structure()
+
+        # lets verify that both are updated
+        id_matches = check_step["primitive"]["id"] == step["primitive"]["id"]
+        if not id_matches:
+            step["primitive"]["id"] = check_step["primitive"]["id"]
+        version_matches = check_step["primitive"]["version"] == step["primitive"]["version"]
+        if not version_matches:
+            step["primitive"]["version"] = check_step["primitive"]["version"]
+        # make sure the digest exists
+        digests_match = "digest" in step["primitive"] and check_step["primitive"]["digest"] == step["primitive"]["digest"]
+        if not digests_match:
+            step["primitive"]["digest"] = check_step["primitive"]["digest"]
```

Comment:
We need to remove digests, not update them. Other teams may update their primitives after we submit our primitives and pipelines. If they do, the digests will change and our submitted pipelines become invalid.

Reply:
Yes, totally right.  I'll remove those lines